pkg/routeagent_driver/handlers/calico/ippool_handler.go

```diff
@@ -0,0 +1,256 @@
+	}
+
+	return nil
+}
+
+func (h *calicoIppoolHandler) RemoteEndpointUpdated(endpoint *submV1.Endpoint) error {
+	err := h.handleRemoteEndpointCreateOrUpdate(endpoint, false)
+	if err != nil {
+		return errors.Wrapf(err, "failed to handle RemoteEndpointUpdated event")
+	}
+
+	return nil
+}
+
+func (h *calicoIppoolHandler) RemoteEndpointRemoved(endpoint *submV1.Endpoint) error {
+	delete(h.remoteEndpoints, endpoint.Name)
+
+	if !h.isGateway {
+		logger.Info("Ignore RemoteEndpointRemoved event (node isn't Gateway)")
+		return nil
+	}
+
+	retErr := error(nil)
+	subnets := extractIPv4Subnets(&endpoint.Spec)
+
+	for _, subnet := range subnets {
+		err := h.deleteIPPool(getEndpointSubnetIPPoolName(endpoint, subnet))
+		if err != nil {
+			logger.Errorf(err, "Error deleting Calico IPPool for ClusterID %q subnet %q", endpoint.Spec.ClusterID, subnet)
+			retErr = err
+		}
+	}
+
+	return retErr
+}
+
+func (h *calicoIppoolHandler) TransitionToNonGateway() error {
+	logger.Info("TransitionToNonGateway")
+
+	h.isGateway = false
+
+	return nil
+}
+
+func (h *calicoIppoolHandler) TransitionToGateway() error {
+	logger.Info("TransitionToGateway")
+
+	retErr := error(nil)
+	h.isGateway = true
+
+	for _, endpoint := range h.remoteEndpoints {
+		err := h.handleRemoteEndpointCreateOrUpdate(endpoint, false)
+		if err != nil {
+			logger.Warningf("Failed to update ippool %s", endpoint.GetName())
+
+			retErr = err
+		}
+	}
+
+	return retErr
+}
+
+func (h *calicoIppoolHandler) Stop(uninstall bool) error {
+	if !uninstall || !h.isGateway {
+		return nil
+	}
+
+	logger.Info("Uninstalling Calico IPPools used for Submariner")
+
+	labelSelector := labels.SelectorFromSet(map[string]string{submarinerIPPool: "true"}).String()
+	err := h.cs.ProjectcalicoV3().IPPools().DeleteCollection(context.TODO(), metav1.DeleteOptions{},
+		metav1.ListOptions{LabelSelector: labelSelector})
+
+	if err != nil && !apierrors.IsNotFound(err) {
+		return errors.Wrapf(err, "Failed to delete Calico IPPools using labelSelector %q", labelSelector)
+	}
+
+	logger.Infof("Successfully delete Calico IPPools using labelSelector %q", labelSelector)
+
+	return nil
+}
+
+func (h *calicoIppoolHandler) handleRemoteEndpointCreateOrUpdate(endpoint *submV1.Endpoint, isCreate bool) error {
+	h.remoteEndpoints[endpoint.Name] = endpoint
+	if !h.isGateway {
+		return nil
+	}
+	var funcToRun createUpdateFunc = h.createIPPool
+	operation := "create"
+
+	if !isCreate {
+		funcToRun = h.updateIPPool
+		operation = "update"
+	}
+
+	subnets := extractIPv4Subnets(&endpoint.Spec)
+	retErr := error(nil)
+
+	for _, subnet := range subnets {
+		iPPoolObj := &calicoapi.IPPool{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:   getEndpointSubnetIPPoolName(endpoint, subnet),
+				Labels: map[string]string{submarinerIPPool: "true"},
+			},
+			Spec: calicoapi.IPPoolSpec{
+				CIDR:        subnet,
+				NATOutgoing: false,
+				Disabled:    true,
+			},
+		}
+
+		err := funcToRun(iPPoolObj)
```

Comment:
I don't think the update operation is appropriate here b/c no field can actually change. The only thing that's variable is the `subnet` but that is also part of the name so it's unique for every instance and can't change. I think what you want to do here is basically "diff" the new subnets with existing `IPPools`, ie 

- obtain the previous `IPPools`
- for each subnet, if there's no previous `IPPool` create it
- delete any previous `IPPools` w/o a corresponding subnet

Reply:
Well, since changing cluster name and CIDRs post submariner deployment isn't supported, we don't need to handle endpoint update event here. I removed the relevant code.